src/cargo/ops/cargo_add/mod.rs

```diff
@@ -104,6 +105,14 @@ pub fn add(workspace: &Workspace<'_>, options: &AddOptions<'_>) -> CargoResult<(
                     options.gctx,
                     &mut registry,
                 )
+                .map_err(|err| {
```

Comment:
Please add a test for this change. Ideally, add it in a commit before, with it passing so it shows the current behavior and this commit would then show how the behavior changed.

Reply:
In principle I agree this should be tested, but the test framework has such high overhead that I think it's unreasonable to use it for such a tiny feature: https://github.com/rust-lang/cargo/issues/15691

Reply:
While there might be overhead, we'd still want to see tests for changes, even ones like this. While there can be improvements, how we currently test is the reality we are currently in and doesn't affect the need for tests.

For example, a test will catch if the error you are matching on changes.

Reply:
I've restructured the code to require an exact error type (so that the integration point is checked by the type system, not a test). Then I've added a localized unit test for the suggestion. I think this way it can test more cases, quicker, and isn't as sensitive to irrelevant UI changes as the end-to-end tests are.